### PR TITLE
Support copy from map to map

### PR DESCRIPTION
**Pull request by vitrun**
resolve issue #44

### CODE REVIEW

File: copier.go [hunk 1]
```diff
@@ -34,6 +34,30 @@ func Copy(toValue interface{}, fromValue interface{}) (err error) {
 		return
 	}
 
+	if fromType.Kind() == reflect.Map && toType.Kind() == reflect.Map {
+		if !fromType.Key().ConvertibleTo(toType.Key()) {
+			return
+		}
+		if to.IsNil() {
+			to.Set(reflect.MakeMapWithSize(toType, from.Len()))
```
Comment (yamil-rivera) [inline copier.go]: First of all, Thank you for the PR! I need this exact functionality and I'm sure I'm not the only one...

Attempting to copy from a type `[]*rawBranchType` to a type `[]branchType` makes this line panic with the error:

`panic: reflect.Set: value of type configs.branchType is not assignable to type []configs.branchType`

where:
`type rawBranchType map[string][]*<some-struct>`
`type branchtype map[string][]<some-struct>`

I think this problem is related with the fact that, in my case, the types on the source are actually pointers to structs/maps that mirror the target structs. (There are pointers to values in the "source" and just values in the "target"). This automatic pointer dereference approach work for copying structs, so should work for copying maps as well. 

The issue might be around the indirection that happens from pointers AND slices types to element types in `func indirectType()` and how those are handled in your code. I'll continue debugging and update this if I find something more concrete.

Comment (vitrun) [reply copier.go:42]: Hi, It works for me:

``` go
type Foo1 struct {
	Name string
	Age  int32
}

type Foo2 struct {
	Name string
}

type MapofSliceofPoint = map[string][]*Foo1

type MapofSliceofStruct = map[string][]Foo2

func TestCopyMapOfSliceOfStruct(t *testing.T) {
	obj1 := MapofSliceofPoint{
		"1": {{Name: "A pure foo, 1.0"}, {Name: "A pure foo, 1.1"}},
		"2": {{Name: "Another pure foo"}},
	}
	obj2 := MapofSliceofStruct{}
	err := copier.Copy(&obj2, obj1)
	if err != nil {
		t.Error("Should not raise error")
	}
	for k, v1 := range obj1 {
		v2, ok := obj2[k]
		if !ok {
			t.Errorf("Map should be copied")
		}
		if len(v1) != len(v2) {
			t.Errorf("Map values should be copied")
		}
		for i := range v1 {
			if v1[i].Name != v2[i].Name {
				t.Errorf("Slice should be copied")
			}
		}
	}
}
```

Comment (yamil-rivera) [reply copier.go:42]: Oh how weird. I'll write and share the test that fails when I find some time. Maybe I'm missing some other special condition that causes the panic in my case.

**Comment by jinzhu** [comment]
Merged, thank you for your PR.

**Pull request closed without merging**